### ISSUE TITLE
Add ECMAScript 6 features

### ISSUE THREAD

**Issue by nzakas** [eslint/espree#10]
This is a rollup that will let us keep track of ES6 progress.

- [x] `let` declarations
- [x] `const` declarations
- [x] Regular expression `u` flag
- [x] Regular expression `y` flag
- [x] Binary literals
- [x] New octal literals
- [x] Arrow functions
- [x] Default parameters
- [x] Rest parameters
- [x] Spread operator
- [x] `super` references
- [x] Classes
- [x] `for-of` loops
- [x] Generators
- [x] Object literal property shorthand
- [x] Object literal method shorthand
- [x] Object literal computed properties
- [x] Template strings
- [x] Destructuring
- [x] Unicode code point escapes
- [x] Allow duplicate object literal properties (except `__proto__`)
- [x] Modules

**Comment by FredKSchott** [comment]
I'm starting work on Unicode code point escapes

**Comment by nzakas** [comment]
Generators and code point escapes have been merged in.

**Comment by nzakas** [comment]
Next priority is template literals. @FredKSchott interested in taking that?

**Comment by FredKSchott** [comment]
definitely, I'll tackle that next (away for the long weekend but I'll take a look next week)

They're supported in V8 now, right? Makes sense to do them next

**Comment by nzakas** [comment]
Yup! And also in io.js now.

I'm planning on attacking arrow functions when I have some energy.

**Comment by xjamundx** [comment]
Interested in taking a shot in default params, though I'm not sure how much work it will be/take I'll have...

**Comment by FredKSchott** [comment]
Template strings are done! I'd be interested in taking on destructuring or rest parameters next. @nzakas thoughts?

**Comment by xjamundx** [comment]
+1 for destructuring.

**Comment by FredKSchott** [comment]
@xjamundx whoops, just saw your post above on taking rest params. Are you still planning on working on that?

**Comment by nzakas** [comment]
Rest params are much easier. Destructuring is tricky. You can decide the level of challenge you want. :)

**Comment by FredKSchott** [comment]
![why not both](http://i0.kym-cdn.com/photos/images/newsfeed/000/538/731/0fc.gif)

**Comment by nzakas** [comment]
:+1:

**Comment by trusktr** [comment]
:+1: for modules and classes!

**Comment by AbrahamAlcaina** [comment]
+1 classes

**Comment by FredKSchott** [comment]
Quick update: I've been completely swamped this week, and haven't been able to work on this at all this week. I probably won't be free again for another week or two. Anyone who wants to grab rest params or destructuring should feel free.

**Comment by koistya** [comment]
:+1:

**Comment by nzakas** [comment]
Thanks for the update, Fred.

**Comment by nzakas** [comment]
Tackling `super` references now.

**Comment by gaearon** [comment]
:+1: to destructuring and arrow functions

**Comment by xjamundx** [comment]
I have like 45 minutes to takea  stab at either. I'll try arrow functions.

**Comment by nzakas** [comment]
I'd be super impressed if you could do arrow functions in 45 minutes, it's a bit complicated because you don't know that an expression is the beginning of an arrow function until the `=>` is encountered.

If you want a nice simple one, rest parameters would probably work.

**Comment by xjamundx** [comment]
Yeah def won't be done in 45, but I think in a few hours I could get this to work. Will open PR shortly... it's coming along. A lot of the core stuff was already ported over for other ES6 stuff...

**Comment by xjamundx** [comment]
Will take a stab at destructuring. Hoping to have a PR with just some broken tests today, but imagine it will take a few weeks to track down every last piece.

**Comment by FredKSchott** [comment]
:+1:

**Comment by xjamundx** [comment]
Hey people this is pretty great. 4 things left!

**Comment by gaearon** [comment]
<3

**Comment by xjamundx** [comment]
:hand: for rest next

**Comment by endel** [comment]
[facebook/esprima](https://github.com/facebook/esprima) supports classes.

**Comment by nzakas** [comment]
Esprima classes support: https://github.com/jquery/esprima/pull/1061

**Comment by xjamundx** [comment]
Well that makes things easier :)

**Comment by nzakas** [comment]
Working on classes now.

**Comment by xjamundx** [comment]
:clap:

**Comment by FredKSchott** [comment]
you guys are killing it :clap: :clap: :clap:

**Comment by nzakas** [comment]
Ran into a bit of trouble with classes. The Harmony branch is different that Esprima's master now (master being the most correct). Will probably take me a bit to work through it unless someone else wants to give it a shot.

**Comment by xjamundx** [comment]
I'm still trying to grok the destructuring defaults spec. Once I'm done with that I'll try to throw my hat into either classes or modules. Whatever's left....

**Comment by niels4** [comment]
Fast work on getting these es6 features implemented!

I noticed an issue with rest parameters though, they don't seem to get added to the scope's variable list. I noticed this when trying to use eslint to validate a function with rest parameters.

```javascript
var func = function (first, ...rest) {
  return rest;
};
```

I am getting a 'rest' is not defined (no-undef) error. The underlying cause seems to be that when the rule's function looks at context.getScope().variables, a variable named 'rest' is not in the list.

**Comment by xjamundx** [comment]
Cool just saw https://github.com/eslint/eslint/issues/1870.

**Comment by nzakas** [comment]
As I mentioned there, this is an issue with escope, not espree.

**Comment by nzakas** [comment]
Don't look now, modules is the last feature!

**Comment by xjamundx** [comment]
![](http://media2.giphy.com/media/kIKDGXcdFGkzm/200.gif)

This is great news!! Well done!

**Comment by fczuardi** [comment]
should we file a separate bug just for the modules support?

In this related bug I posted an example of how espree differs from esprima when used with escope https://github.com/eslint/eslint/issues/1978#issuecomment-77903737

**Comment by nzakas** [comment]
Modules are already properly supported.
